# DATA QUALITY ASSESSMENT IN R
    

**Grace Ndunda**

**12/8/2022**

### Task: Draft a document to the client identifying the data quality issues and strategies to mitigate these issues. Refer to ‘Data Quality Framework Table’ and resources below for criteria and dimensions which you should consider.

In [ ]:
%load_ext rpy2.ipython

The rpy2.ipython extension is already loaded. To reload it, use:
  %reload_ext rpy2.ipython


In [ ]:

!pip uninstall rpy2 -y
!pip install rpy2==3.5.1


Found existing installation: rpy2 3.5.1
Uninstalling rpy2-3.5.1:
  Successfully uninstalled rpy2-3.5.1
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached rpy2-3.5.1-cp38-cp38-linux_x86_64.whl


In [ ]:
%%R
install.packages("readxl")

(as ‘lib’ is unspecified)







	‘/tmp/Rtmph3fxH0/downloaded_packages’



In [ ]:
%%R
library("readxl")

In [ ]:
%%R
library(ggplot2) 
library(tidyverse)

In [ ]:
%%R
Transactions<- read_excel("/content/KPMG_VI_New_raw_data_update_final .xlsx", 
                                                sheet = "Transactions", skip = 1)
CustomerList <- read_excel("/content/KPMG_VI_New_raw_data_update_final .xlsx", 
    sheet = "NewCustomerList", skip = 1)

New names:
• `` -> `...17`
• `` -> `...18`
• `` -> `...19`
• `` -> `...20`
• `` -> `...21`


In [ ]:
%%R
CustomerDemographic <- read_excel("/content/KPMG_VI_New_raw_data_update_final .xlsx", 
    sheet = "CustomerDemographic", skip = 1)

CustomerAddress <- read_excel("/content/KPMG_VI_New_raw_data_update_final .xlsx", 
    sheet = "CustomerAddress", skip = 1)

# Data quality Dimensions

## data quality dimensions include;
 

*   Accuracy
*   completeness


*   Consistency
*   Relevance


*   Validity
*   Uniqueness








## Accuracy

**This deals with how information is accurate concerning a certain variable**

# Transactions

In [ ]:
%%R
summary(Transactions)

 transaction_id    product_id      customer_id    
 Min.   :    1   Min.   :  0.00   Min.   :   1.0  
 1st Qu.: 5001   1st Qu.: 18.00   1st Qu.: 857.8  
 Median :10000   Median : 44.00   Median :1736.0  
 Mean   :10000   Mean   : 45.36   Mean   :1738.2  
 3rd Qu.:15000   3rd Qu.: 72.00   3rd Qu.:2613.0  
 Max.   :20000   Max.   :100.00   Max.   :5034.0  
                                                  
 transaction_date                 online_order    order_status      
 Min.   :2017-01-01 00:00:00.00   Mode :logical   Length:20000      
 1st Qu.:2017-04-01 00:00:00.00   FALSE:9811      Class :character  
 Median :2017-07-03 00:00:00.00   TRUE :9829      Mode  :character  
 Mean   :2017-07-01 14:08:05.27   NA's :360                         
 3rd Qu.:2017-10-02 00:00:00.00                                     
 Max.   :2017-12-30 00:00:00.00                                     
                                                                    
    brand           product_line       p

**Everything seems to be alright with the type of data of the variables, minimum values, maximum values, and format. Therefore, I assume the accuracy of the datasets given is high.**

# Customer List

In [ ]:
%%R
summary(CustomerList)

  first_name         last_name            gender         
 Length:1000        Length:1000        Length:1000       
 Class :character   Class :character   Class :character  
 Mode  :character   Mode  :character   Mode  :character  
                                                         
                                                         
                                                         
 past_3_years_bike_related_purchases     DOB             job_title        
 Length:1000                         Length:1000        Length:1000       
 Class :character                    Class :character   Class :character  
 Mode  :character                    Mode  :character   Mode  :character  
                                                                          
                                                                          
                                                                          
 job_industry_category wealth_segment     deceased_indicator   owns_c

**I see problems with the purchases on bike related purchases, DOB, postcode, and property valuation for having a wrong format. Additionally there is a lack of customer ID, which makes validation of information more difficult between datasets.**

# Customer Demographics

In [ ]:
%%R
summary(CustomerDemographic)

  customer_id    first_name         last_name            gender         
 Min.   :   1   Length:4000        Length:4000        Length:4000       
 1st Qu.:1001   Class :character   Class :character   Class :character  
 Median :2000   Mode  :character   Mode  :character   Mode  :character  
 Mean   :2000                                                           
 3rd Qu.:3000                                                           
 Max.   :4000                                                           
                                                                        
 past_3_years_bike_related_purchases     DOB             job_title        
 Min.   : 0.00                       Length:4000        Length:4000       
 1st Qu.:24.00                       Class :character   Class :character  
 Median :48.00                       Mode  :character   Mode  :character  
 Mean   :48.89                                                            
 3rd Qu.:73.00                           

**I see problems with, DOB for having a wrong format**

# Customer Adress

In [ ]:
%%R
summary(CustomerAddress)

  customer_id     address             postcode       state          
 Min.   :   1   Length:3999        Min.   :2000   Length:3999       
 1st Qu.:1004   Class :character   1st Qu.:2200   Class :character  
 Median :2004   Mode  :character   Median :2768   Mode  :character  
 Mean   :2004                      Mean   :2986                     
 3rd Qu.:3004                      3rd Qu.:3750                     
 Max.   :4003                      Max.   :4883                     
   country          property_valuation
 Length:3999        Min.   : 1.000    
 Class :character   1st Qu.: 6.000    
 Mode  :character   Median : 8.000    
                    Mean   : 7.514    
                    3rd Qu.:10.000    
                    Max.   :12.000    


**Everything seems to be alright with the type of data of the variables, minimum values, maximum values, and format. Therefore, we assume the accuracy of the datasets given is high.**

## Completeness

**Analyzing the missing values in the dataset**

# Transactions

In [ ]:
%%R
sum(is.na(Transactions))


[1] 1542


In [ ]:
#checking for null value
%%R
sum(is.na(Transactions))
# calculating the product of dimensions of dataframe 
totalcells = prod(dim(Transactions))
print("Total number of cells ")
print(totalcells)
  
# calculating the number of cells with na
missingcells = sum(is.na(Transactions))
print("Missing value cells")
print(missingcells)
  
# calculating percentage of missing values
percentage = (missingcells * 100 )/(totalcells)
print("Percentage of missing values' cells")
print (percentage)

[1] "Total number of cells "
[1] 260000
[1] "Missing value cells"
[1] 1542
[1] "Percentage of missing values' cells"
[1] 0.5930769


**From the above analysis the number of missing value cells is 1542 which is about 0.59% of the entire dataset.**

## Customer List

In [ ]:
#checking for null value
%%R
sum(is.na(CustomerList))
# calculating the product of dimensions of dataframe 
totalcells = prod(dim(CustomerList))
print("Total number of cells ")
print(totalcells)
  
# calculating the number of cells with na
missingcells = sum(is.na(CustomerList))
print("Missing value cells")
print(missingcells)
  
# calculating percentage of missing values
percentage = (missingcells * 100 )/(totalcells)
print("Percentage of missing values' cells")
print (percentage)

[1] "Total number of cells "
[1] 23000
[1] "Missing value cells"
[1] 152
[1] "Percentage of missing values' cells"
[1] 0.6608696


**The above analysis shows a 0.66% of missing values. However, this variable might not be critical for analysis**

# Customer Demographics

In [ ]:
#checking for null value
%%R
sum(is.na(CustomerDemographic))
# calculating the product of dimensions of dataframe 
totalcells = prod(dim(CustomerDemographic))
print("Total number of cells ")
print(totalcells)
  
# calculating the number of cells with na
missingcells = sum(is.na(CustomerDemographic))
print("Missing value cells")
print(missingcells)
  
# calculating percentage of missing values
percentage = (missingcells * 100 )/(totalcells)
print("Percentage of missing values' cells")
print (percentage)

[1] "Total number of cells "
[1] 52000
[1] "Missing value cells"
[1] 1045
[1] "Percentage of missing values' cells"
[1] 2.009615


**The 5 variables displayed above show missing values, the job_title variable shows a significant missing values; however, this variable might not be critical to the analysis. On the other hand, the other two variables show signifficant completiness.**

# Customer Address

In [ ]:
#checking for null value
%%R
sum(is.na(CustomerAddress))
# calculating the product of dimensions of dataframe 
totalcells = prod(dim(CustomerAddress))
print("Total number of cells ")
print(totalcells)
  
# calculating the number of cells with na
missingcells = sum(is.na(CustomerAddress))
print("Missing value cells")
print(missingcells)
  
# calculating percentage of missing values
percentage = (missingcells * 100 )/(totalcells)
print("Percentage of missing values' cells")
print (percentage)

[1] "Total number of cells "
[1] 23994
[1] "Missing value cells"
[1] 0
[1] "Percentage of missing values' cells"
[1] 0


**This means that there are not missing values, therefore the whole dataset is complete.**

# Consistency

Consistency refers to having the same data across different datasets.

First, we identify the name of the variables that are repeated accross datasets.

In [ ]:
%%R
colnames(Transactions)

 [1] "transaction_id"          "product_id"             
 [3] "customer_id"             "transaction_date"       
 [5] "online_order"            "order_status"           
 [7] "brand"                   "product_line"           
 [9] "product_class"           "product_size"           
[11] "list_price"              "standard_cost"          
[13] "product_first_sold_date"


In [ ]:
%%R
colnames(CustomerList)

 [1] "first_name"                          "last_name"                          
 [3] "gender"                              "past_3_years_bike_related_purchases"
 [5] "DOB"                                 "job_title"                          
 [7] "job_industry_category"               "wealth_segment"                     
 [9] "deceased_indicator"                  "owns_car"                           
[11] "tenure"                              "address"                            
[13] "postcode"                            "state"                              
[15] "country"                             "property_valuation"                 
[17] "...17"                               "...18"                              
[19] "...19"                               "...20"                              
[21] "...21"                               "Rank"                               
[23] "Value"                              


In [ ]:
%%R
colnames(CustomerDemographic)

 [1] "customer_id"                         "first_name"                         
 [3] "last_name"                           "gender"                             
 [5] "past_3_years_bike_related_purchases" "DOB"                                
 [7] "job_title"                           "job_industry_category"              
 [9] "wealth_segment"                      "deceased_indicator"                 
[11] "default"                             "owns_car"                           
[13] "tenure"                             


In [ ]:
%%R
colnames(CustomerAddress)

[1] "customer_id"        "address"            "postcode"          
[4] "state"              "country"            "property_valuation"


**At a first glance it is possible to determine the shared variables between datasets. In particular, Customer List, seems to include most of the variables in Customer Demographics and Customer Address.**

**There are 11 variables that are shared**

# Timeliness

**There is no time stamp that shows how the data is managed and the availability of it. Therefore a timeliness parameter cannot be analyzed.**

# Relevancy

**Some of the variables are not that relevant to some of the datasets. For example the variable default has no relevance in the CustomerDemographic dataset. In the same way, the duplication of data in the CustomerList makes the demographics included there not relevant. Additionally, a CustomerID is missing from this dataset.**

# Validity

**As seen in the accuracy section some variables do not follow an expected format that they are supposed to, therefore the data is not validated and standardized for these variables.**

# Uniqueness

**The following sections show the uniqueness value of each variable per dataset.**

### Transactions

In [61]:
%%R
df_uniq =unique(Transactions)
length(df_uniq)
   

[1] 13


### Customer List

In [63]:
%%R
df_uniqlist =unique(CustomerList)
length(df_uniqlist)

[1] 23


### Customer Address

In [65]:
%%R
df_uniqadress =unique(CustomerAddress)
length(df_uniqadress)

[1] 6


### Customer Demographic

In [66]:
%%R
df_uniqdemo =unique(CustomerDemographic)
length(df_uniqdemo)

[1] 13


# **Overall Changes**

After analyzing each of the different categories I can determine the level of the data dimension into three cathegories. Improve, will mean that there is a lot to improve, and it is a priority to solve before moving into the next stage. Minor, meaning that it is fairly good and minor changes should be made; and Okay, meaning that no issues could be found and its ready for the next stage.

Accuracy: minor changes to be made

Completeness: Minor changes to be made

Consistency: Improve before next stage

Timeliness: Improve before next stage

Relevancy: Minor changes to be made

Validity: Minor changes to be made

Uniqueness: Okay for next stage

# Mitigation measures

Accuracy: As of right now the accuracy of the data is relatively high. Naturally some cleanliness in the data and standardization will increase this accuracy.

Completeness: Most of the data is complete, the factor of missing values is usually non significants. Some of the variables that are not complete, are duplicated with other dataset, so after validating the entrees, the completeness level will increase even more. Additionally, it is necessary to the source of the missing data, if they ar missing at random, they can be ignored for future analysis. If not, they can be predicted with the mean or meadia of the categories missing.

Consistency: This is the main problem of the data. A lot of data is not consistent across datasets. It is recommended to add a Customer ID for this matter in the customer list. There is a significant number of entries that differ across datasets. Major cleaning must be performed.

Timeliness: Adding a time stamp would create the timeliness parameter that will allow the company to know when the data is available.

Relevancy: Besides the variable default in the CustomerDemographics dataset, all the variables seem to be relevant.

Validity: Some minor adjustments must be done. Specially on the variables regarding dates. A standardized format must be implemented for future datasets.

Uniqueness: The data seems to have unique entries, however, a better control can be added once all the datasets are dependent on the CustomerID